pkg/genericapiserver/default_storage_factory_builder.go

```diff
@@ -0,0 +1,134 @@
+package genericapiserver
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	"k8s.io/kubernetes/pkg/runtime"
+	etcdstorage "k8s.io/kubernetes/pkg/storage/etcd"
+	"k8s.io/kubernetes/pkg/util/config"
+)
+
+type DefaultStorageFactoryBuilder struct {
+	runtimeConfig                  config.ConfigurationMap
+	etcdConfig                     etcdstorage.EtcdConfig
+	serializer                     runtime.NegotiatedSerializer
+	storageGroupsToEncodingVersion map[string]unversioned.GroupVersion
+	defaultResourceEncoding        *DefaultResourceEncodingConfig
+	defaultResourceConfig          *ResourceConfig
+}
+
+func NewDefaultStorageFactoryBuilder(rc config.ConfigurationMap, ec etcdstorage.EtcdConfig, s runtime.NegotiatedSerializer,
+	ev map[string]unversioned.GroupVersion, encoding *DefaultResourceEncodingConfig, config *ResourceConfig) *DefaultStorageFactoryBuilder {
+	return &DefaultStorageFactoryBuilder{
+		runtimeConfig:                  rc,
+		etcdConfig:                     ec,
+		serializer:                     s,
+		storageGroupsToEncodingVersion: ev,
+		defaultResourceEncoding:        encoding,
+		defaultResourceConfig:          config,
+	}
+}
+
+func (b *DefaultStorageFactoryBuilder) NewFactory() (*DefaultStorageFactory, error) {
+	apiResourceConfigSource, err := b.parseRuntimeConfig()
+	if err != nil {
+		return nil, fmt.Errorf("error in parsing runtime-config: %s", err)
+	}
+
+	resourceEncoding := b.defaultResourceEncoding
+	for group, storageEncodingVersion := range b.storageGroupsToEncodingVersion {
+		resourceEncoding.SetVersionEncoding(group, storageEncodingVersion, unversioned.GroupVersion{Group: group, Version: runtime.APIVersionInternal})
+	}
+
+	return NewDefaultStorageFactory(b.etcdConfig, b.serializer, resourceEncoding, apiResourceConfigSource), nil
+}
+
+// Parses the given runtime-config and formats it into APIResourceConfigSource
+func (b *DefaultStorageFactoryBuilder) parseRuntimeConfig() (APIResourceConfigSource, error) {
```

Comment:
Could we factor this to take the `runtimeConfig` and return the `APIResourceConfigSource`?  If we did that, then we'd have a few helpers to go from CLI-ish input to interfaces we need for the `DefaultStorageFactory` without needing a builder on top.  We'd lose the mutability that this current factorization gives us, but I'm not sure that's a thing we really want to take advantage of.